### PR TITLE
test: fix flaky GoAwayThenClose

### LINKED ISSUE #3253: Flaky test: GoAwayThenClose
From the most recent cron run, although I've seen it at least one other time this week:
```
--- FAIL: Test (159.44s)
    --- FAIL: Test/GoAwayThenClose (20.00s)
        end2end_test.go:7160: UnaryCall(_) = _, rpc error: code = DeadlineExceeded desc = context deadline exceeded; want _, nil
FAIL
```

### PR DESCRIPTION
**Pull request by menghanl**
In this `GoAwayThenClose` test, we
1. make a streaming RPC on a connection
1. graceful stop the connection from the server, to send a GOAWAY
1. hard stop the server, so the client will create a new connection to another server
1. wait for the streaming RPC to fail
1. make unary RPC and expect them to connect to the second server

Before this fix, 2 and 3 can happen too soon, so the RPC in 1 would fail
and then transparent retry (because the stream is unprocessed by the
server in that case).

This retry attempt could pick the new connection, and then the RPC would
block until timeout.
After this streaming RPC fails, we make unary RPCs with the same
deadline (**note:** deadline not timeout) as the streaming RPC and expect
them to succeed. But they will also fail due to timeout.

The fix is to make a round-trip on the streaming RPC first, to make sure
it actually goes on the first connection.

fixes #3253

### CODE REVIEW

**Comment by menghanl** [comment]
No failure in 50000 runs after the fix.

File: test/end2end_test.go [hunk 1]
```diff
@@ -6954,12 +6958,19 @@ func (s) TestGoAwayThenClose(t *testing.T) {
...
+	// We use a long-lived RPC because it will cause GracefulStop to send
+	// GO_AWAY, but the connection doesn't get closed until the server stops and
+	// the client receives the error.
 	stream, err := client.FullDuplexCall(ctx)
 	if err != nil {
 		t.Fatalf("FullDuplexCall(_) = _, %v; want _, nil", err)
 	}
+	if _, err = stream.Recv(); err != nil {
+		t.Fatalf("unexpected error from first recv: %v", err)
+	}
```
Comment (dfawley) [inline test/end2end_test.go:6973]: Somewhat related... Do we have another test that will do sends/receives/finish RPCs on a gracefully-stopped-but-not-fully-stopped server?  If not, then adding another `Recv` after `conn2Established` but before `s1.Stop()` would be a cheap way of testing it.

**Review by dfawley — APPROVED** [review]
Wow, nice work!  Seems like a tricky one to track down.

File: test/end2end_test.go (hunk 1 above)
Comment (menghanl) [reply test/end2end_test.go:6973]: That's covered in the other GoAway tests: https://github.com/grpc/grpc-go/blob/master/test/end2end_test.go#L1114

**Pull request merged**